docs/site/migration/components/project-layout.md

```diff
@@ -0,0 +1,131 @@
+content:
+
+```ts
+module.exports = function initializeComponent(loopbackApplication, options) {
+  // implementation
+};
+```
+
+A component is typically added to a LB3 application by creating a new entry in
+`server/component-config.json`, see
+[LoopBack components](https://loopback.io/doc/en/lb3/LoopBack-components.html).
+
+For example:
+
+```json
+{
+  "loopback-explorer": {
+    "mountPath": "/explorer"
+  }
+}
+```
+
+This allows the component to receive configuration. App developers can provide
+environment-specific configuration by using `component-config.{env}.json` files.
+
+## LoopBack 4 layout
+
+As explained in [Creating components](../../Creating-Components.md) and
+[Using components](../../Using-components.md), a typical LoopBack 4 component is
+an npm package exporting a Component class.
+
+The component class is usually implemented inside
+`src/{component-name}.component.ts` file, for example
+[src/metrics.component.ts](https://github.com/strongloop/loopback-next/blob/38f10b240551227d2d030c2fe8ee206880c9e029/extensions/metrics/src/metrics.component.ts):
+
+```ts
+import {Application, Component, CoreBindings} from '@loopback/core';
+import {bind, config, ContextTags, inject} from '@loopback/context';
+import {MetricsBindings} from './keys';
+import {DEFAULT_METRICS_OPTIONS, MetricsOptions} from './types';
+
+@bind({tags: {[ContextTags.KEY]: MetricsBindings.COMPONENT}})
+export class MetricsComponent implements Component {
+  constructor(
+    @inject(CoreBindings.APPLICATION_INSTANCE)
+    private application: Application,
+    @config()
+    options: MetricsOptions = DEFAULT_METRICS_OPTIONS,
+  ) {
+    // ...
+  }
+  // ...
+}
+```
+
+The code snippet above also shows how a LoopBack 4 component can receive
+configuration and the target application object.
+
+## Usage instructions
+
+LoopBack 4 components are added to applications inside application constructor.
+
+First, the application file needs to import the component class:
+
+```ts
+import {MetricsComponent} from '@loopback/extension-metrics';
+```
+
+Then in the constructor, add the component to your application:
+
+```ts
+this.component(MetricsComponent);
+```
+
+Finally, the application can configure the component by adding the following
+code to its constructor:
+
+```ts
+this.configure(MetricsBindings.COMPONENT).to({
+  // the configuration
+});
+```
+
+## Migration steps
+
+It is not feasible to migrate a LoopBack 3 component project to a LoopBack 4
+component project in a series of incremental changes done within the same
+repository. We recommend to create a new project using
+[`lb4 extension`](../../Extension-generator.md) CLI command and then
+incrementally migrate artifacts from the original LoopBack 3 component to the
+new project.
+
+<!--
+FIXME: create a follow-up task to update Extension template to follow
+the latest style as used by `extensions/metrics`:
+- inject the target app and component config to the component class,
```

Comment:
It's also possible to build a component booter to load declarative json files similar as LB3. I plan to contribute a `Configuration` booter to load `json/js/yaml/rc` files from `configs` directory and use `app.configure` to bind such entries.

Reply:
Sure. Such booter is out of scope of my pull request though. First, it's not implemented yet. Second, this FIXME comment is just a placeholder, the changes are tracked by the following GitHub issue I created few moments ago: https://github.com/strongloop/loopback-next/issues/5336